### ISSUE TITLE
How to pass a Go function pointer/callback to a C function?

### ISSUE THREAD

**Issue by delthas** [xlab/c-for-go#55]
Hello,

Thanks for creating this great tool. I'm creating a binding for [Soundpipe](https://github.com/PaulBatchelor/Soundpipe). I was just wondering, the library has a function called:
```int sp_process_plot(sp_data *sp, void *ud, void (*callback)(sp_data *, void *)) { ... }```
Which gives me this Go signature after running c-for-go:
```func Process(sp *Data, ud unsafe.Pointer, callback *func(arg0 *Data, arg1 unsafe.Pointer)) int32 { ... }```
So, I have created a Go callback function like this:
```func writeOsc(sp *soundpipe.Data, data unsafe.Pointer) { ... }```
Then I want to pass `writeOsc` as callback to `Process`:
```soundpipe.Process(sp, &ud, writeOsc)```
But the types do *not* match for the callback:
- my writeOsc is a `type func (sp *soundpipe.Data, data unsafe.Pointer)`
- the callback is a `type *func (sp *soundpipe.Data, data unsafe.Pointer)`   (notice the additional *)

I tried to do `soundpipe.Process(sp, &ud, &writeOsc`, but Go says I can't take the address of `writeOsc`. Any clue on how I should pass a Go function callback to this function?

Thanks.

**Comment by faisal00813** [comment]
@delthas Any luck with this issue? I am trying to solve a similar one.

**Comment by xlab** [comment]
@faisal00813 I think that creating a bridge function is the only way around. A function that calls some other function by pointer, but itself having a symbol declaration, so the generator could take it.

E.g. 
https://github.com/vulkan-go/vulkan/blob/master/vk_bridge.c
https://github.com/vulkan-go/vulkan/blob/master/vulkan.yml#L43-L44

**Comment by xlab** [comment]
@delthas @faisal00813 hey, isn't this an example of what is needed?
https://github.com/xlab/vorbis-go/blob/master/cmd/vorbis-player/main.go#L134

**Comment by faisal00813** [comment]
@xlab Thank you for pointing to that code. It helped.
But my issue was to generate functions in GO from the  inline callback functions. e.g https://github.com/hyperledger/indy-sdk/blob/4b2cc8ddf83a06dc375fee89bdd2a878beab706e/vcx/libvcx/include/vcx.h#L140
I think this is not supported as of now.

Workaround was to define the callbacks and then use them in function. e.g https://github.com/faisal00813/vcx-go/blob/f70b84cdeb83f17c52f0c72a04ddff65fc49937a/vcx/vcx.h#L42

**Comment by xlab** [comment]
@faisal00813 yep, you should have a type, btw, I advice to avoid modifying the original headers, and instead add another .h to the YAML spec with custom code and bridges. Good luck
